## Below, we will be using Python pandas to import and manipulate data from a bike rental company (Citibike) and weather data (Newark airport) to see correlations between weather and bike rentals!

### First, we will import Pandas and Glob, we then import the relevant CSV's, we also import create_engine from sqlalchemy

In [137]:
import pandas as pd
import glob
from sqlalchemy import create_engine
!pip install psycopg2

## We will now connect to our database using SQLAlchemy, we also test the connection to ensure it is successful.

In [138]:

DATABASE_URL = 'postgresql://postgres:Anglll5567@localhost:5432/Bike Rental'

try:
    
    engine = create_engine(DATABASE_URL)
    
    
    with engine.connect() as connection:
        print("Connection to the database is successful.")
        
except Exception as e:
    print(f"An error occurred: {e}")

Connection to the database is successful.


In [139]:
bike_file_paths = [r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201601-citibike-tripdata.csv",
              r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201602-citibike-tripdata.csv",
              r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201603-citibike-tripdata.csv",
              r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201604-citibike-tripdata.csv",
              r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201605-citibike-tripdata.csv",
              r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201606-citibike-tripdata.csv",
              r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201607-citibike-tripdata.csv",
              r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201608-citibike-tripdata.csv",
              r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201609-citibike-tripdata.csv",
              r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201610-citibike-tripdata.csv",
              r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201611-citibike-tripdata.csv",
              r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\JC-201612-citibike-tripdata.csv"]

In [140]:
weather_file_paths = r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\newark_airport_2016.csv"

## We now create DataFrames for bike rentals and weather report, and show the head of these.

In [141]:
weather_df = pd.read_csv(r"C:\Users\maxpe\OneDrive\Documents\Codecademy\Code Examples\Rent A Bike Python and SQL\bike-rental-starter-kit\data\newark_airport_2016.csv")
weather_df.head()

STATION                                         NAME        DATE  \
0  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-01   
1  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-02   
2  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-03   
3  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-04   
4  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-05   

    AWND  PGTM  PRCP  SNOW  SNWD  TAVG  TMAX  TMIN  TSUN  WDF2   WDF5  WSF2  \
0  12.75   NaN   0.0   0.0   0.0    41    43    34   NaN   270  280.0  25.9   
1   9.40   NaN   0.0   0.0   0.0    36    42    30   NaN   260  260.0  21.0   
2  10.29   NaN   0.0   0.0   0.0    37    47    28   NaN   270  250.0  23.9   
3  17.22   NaN   0.0   0.0   0.0    32    35    14   NaN   330  330.0  25.9   
4   9.84   NaN   0.0   0.0   0.0    19    31    10   NaN   360  350.0  25.1   

   WSF5  
0  35.1  
1  25.1  
2  30.0  
3  33.1  
4  31.1

### Number of rows in Dataframe

In [142]:
num_rows_weather = len(weather_df)
num_rows_weather

366

In [143]:
bike_rental_dfs = {}
for file_path in bike_file_paths:
    file_name = file_path.split('\\')[-1]  
    bike_rental_dfs[file_name] = pd.read_csv(file_path)

combined_df = pd.concat(bike_rental_dfs.values(), ignore_index=True)
combined_df.head()

Trip Duration           Start Time            Stop Time  Start Station ID  \
0            362  2016-01-01 00:02:52  2016-01-01 00:08:54              3186   
1            200  2016-01-01 00:18:22  2016-01-01 00:21:42              3186   
2            202  2016-01-01 00:18:25  2016-01-01 00:21:47              3186   
3            248  2016-01-01 00:23:13  2016-01-01 00:27:21              3209   
4            903  2016-01-01 01:03:20  2016-01-01 01:18:24              3195   

  Start Station Name  Start Station Latitude  Start Station Longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   

   End Station ID End Station Name  End Station Latitude  \
0            3209     Brunswick St             40.724176   
1            3213   Van Vorst Park             40.718489   
2            3213   Van Vorst Park             40.718489   
3            3203    Hamilton Park             40.727596   
4            3210   Pershing Field             40.742677   

   End Station Longitude  Bike ID   User Type  Birth Year  Gender  
0             -74.050656    24647  Subscriber      1964.0       2  
1             -74.047727    24605  Subscriber      1962.0       1  
2             -74.047727    24689  Subscriber      1962.0       2  
3             -74.044247    24693  Subscriber      1984.0       1  
4             -74.051789    24573    Customer         NaN       0

### Number of rows in Dataframe

In [144]:
num_rows = len(combined_df)
num_rows

247584

### Use .info() to see DataFrame information, we know there are 247584 rows. This means we can see some rows have a Null value, we can also see some Data Types are incorrect and need updating.

In [145]:
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 247584 entries, 0 to 247583
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Trip Duration            247584 non-null  int64  
 1   Start Time               247584 non-null  object 
 2   Stop Time                247584 non-null  object 
 3   Start Station ID         247584 non-null  int64  
 4   Start Station Name       247584 non-null  object 
 5   Start Station Latitude   247584 non-null  float64
 6   Start Station Longitude  247584 non-null  float64
 7   End Station ID           247584 non-null  int64  
 8   End Station Name         247584 non-null  object 
 9   End Station Latitude     247584 non-null  float64
 10  End Station Longitude    247584 non-null  float64
 11  Bike ID                  247584 non-null  int64  
 12  User Type                247204 non-null  object 
 13  Birth Year               228585 non-null  float64
 14  Gend

In [146]:
combined_df['Start Time'] = pd.to_datetime(combined_df['Start Time'])
combined_df['Stop Time'] = pd.to_datetime(combined_df['Stop Time'])

### We check the DataFrames for duplicated data, no True values occur so this indicates no duplicate data.

In [147]:
combined_df.duplicated()



0         False
1         False
2         False
3         False
4         False
          ...  
247579    False
247580    False
247581    False
247582    False
247583    False
Length: 247584, dtype: bool

### We use .describe(), we can see some missing values in Birth Year, a long max trip time, and also a low Birth Year of 1920

In [148]:
combined_df.describe()

Trip Duration                     Start Time  \
count   2.475840e+05                         247584   
mean    8.856305e+02  2016-07-29 05:55:07.541335040   
min     6.100000e+01            2016-01-01 00:02:52   
25%     2.480000e+02            2016-05-27 07:46:06   
50%     3.900000e+02            2016-08-10 09:23:50   
75%     6.660000e+02     2016-10-05 17:25:05.500000   
max     1.632981e+07            2016-12-31 23:44:50   
std     3.593798e+04                            NaN   

                           Stop Time  Start Station ID  \
count                         247584     247584.000000   
mean   2016-07-29 06:09:53.671073536       3207.065206   
min              2016-01-01 00:08:54       3183.000000   
25%    2016-05-27 07:54:40.249999872       3186.000000   
50%       2016-08-10 09:34:32.500000       3201.000000   
75%    2016-10-05 17:33:00.750000128       3211.000000   
max              2017-01-18 14:26:46       3426.000000   
std                              NaN         26.955103   

       Start Station Latitude  Start Station Longitude  End Station ID  \
count           247584.000000            247584.000000   247584.000000   
mean                40.723121               -74.046438     3203.572553   
min                 40.692640               -74.096937      147.000000   
25%                 40.717732               -74.050656     3186.000000   
50%                 40.721525               -74.044247     3199.000000   
75%                 40.727596               -74.038051     3211.000000   
max                 40.752559               -74.032108     3426.000000   
std                  0.008199                 0.011211       61.579494   

       End Station Latitude  End Station Longitude        Bike ID  \
count         247584.000000          247584.000000  247584.000000   
mean              40.722594             -74.045855   24935.260481   
min               40.692216             -74.096937   14552.000000   
25%               40.716540             -74.050444   24491.000000   
50%               40.721124             -74.043117   24609.000000   
75%               40.727224             -74.036486   24719.000000   
max               40.801343             -73.957390   27274.000000   
std                0.007958               0.011283     748.469712   

          Birth Year         Gender  
count  228585.000000  247584.000000  
mean     1979.335276       1.123534  
min      1900.000000       0.000000  
25%      1974.000000       1.000000  
50%      1981.000000       1.000000  
75%      1986.000000       1.000000  
max      2000.000000       2.000000  
std         9.596809       0.518687

### We check the longest trip durations, we notice one trip is much longer than any other. We wont remove this, however, we will note this because it could be a potential problem.

In [149]:
top_10_longest_trips = combined_df.sort_values(by='Trip Duration', ascending=False).head(10)
top_10_longest_trips

Trip Duration          Start Time           Stop Time  \
24632        16329808 2016-03-22 07:02:10 2016-09-27 07:05:38   
229148        4826890 2016-11-23 17:38:36 2017-01-18 14:26:46   
10712         2104123 2016-02-12 07:27:56 2016-03-07 15:56:40   
10748         2100551 2016-02-12 08:31:06 2016-03-07 16:00:18   
10850         2071209 2016-02-12 16:32:54 2016-03-07 15:53:03   
43986         1837255 2016-04-28 09:05:14 2016-05-19 15:26:09   
2435          1569765 2016-01-09 05:49:39 2016-01-27 09:52:25   
158958        1532001 2016-09-11 16:32:21 2016-09-29 10:05:42   
141628        1258736 2016-08-26 23:19:01 2016-09-10 12:57:58   
150496        1120971 2016-09-03 22:05:27 2016-09-16 21:28:18   

        Start Station ID    Start Station Name  Start Station Latitude  \
24632               3215           Central Ave               40.746730   
229148              3280           Astor Place               40.719282   
10712               3214      Essex Light Rail               40.712774   
10748               3184           Paulus Hook               40.714145   
10850               3183        Exchange Place               40.716247   
43986               3192    Liberty Light Rail               40.711242   
2435                3197              North St               40.752559   
158958              3197              North St               40.752559   
141628              3198      Heights Elevator               40.748716   
150496              3189  West Side Light Rail               40.714402   

        Start Station Longitude  End Station ID      End Station Name  \
24632                -74.049251            3267          Morris Canal   
229148               -74.071262            3426            JCBS Depot   
10712                -74.036486            3183        Exchange Place   
10748                -74.033552            3183        Exchange Place   
10850                -74.033459            3183        Exchange Place   
43986                -74.055701            3183        Exchange Place   
2435                 -74.044725            3201                Dey St   
158958               -74.044725            3210        Pershing Field   
141628               -74.040443            3197              North St   
150496               -74.088772            3189  West Side Light Rail   

        End Station Latitude  End Station Longitude  Bike ID   User Type  \
24632              40.712419             -74.038526    24519  Subscriber   
229148             40.709651             -74.068601    24711  Subscriber   
10712              40.716247             -74.033459    24555  Subscriber   
10748              40.716247             -74.033459    24720  Subscriber   
10850              40.716247             -74.033459    24664  Subscriber   
43986              40.716247             -74.033459    24515  Subscriber   
2435               40.737711             -74.066921    24629    Customer   
158958             40.742677             -74.051789    24522  Subscriber   
141628             40.752559             -74.044725    26192    Customer   
150496             40.714402             -74.088772    24556  Subscriber   

        Birth Year  Gender  
24632       1975.0       1  
229148      1989.0       1  
10712       1968.0       1  
10748       1970.0       1  
10850       1988.0       2  
43986       1964.0       1  
2435           NaN       0  
158958      1991.0       1  
141628         NaN       0  
150496      1970.0       1

In [150]:
missing_birth_year_data = combined_df[combined_df['Birth Year'].isna()]
missing_birth_year_data

Trip Duration          Start Time           Stop Time  \
4                 903 2016-01-01 01:03:20 2016-01-01 01:18:24   
5                 883 2016-01-01 01:03:28 2016-01-01 01:18:11   
22                988 2016-01-01 03:16:33 2016-01-01 03:33:02   
53               3090 2016-01-01 11:07:15 2016-01-01 11:58:46   
57                788 2016-01-01 11:50:30 2016-01-01 12:03:39   
...               ...                 ...                 ...   
247442           2484 2016-12-31 14:35:19 2016-12-31 15:16:44   
247443          79669 2016-12-31 14:35:25 2017-01-01 12:43:14   
247537            224 2016-12-31 18:01:38 2016-12-31 18:05:23   
247561            440 2016-12-31 20:12:26 2016-12-31 20:19:47   
247580           2749 2016-12-31 23:29:39 2017-01-01 00:15:29   

        Start Station ID Start Station Name  Start Station Latitude  \
4                   3195            Sip Ave               40.730743   
5                   3195            Sip Ave               40.730743   
22                  3196     Riverview Park               40.744319   
53                  3203      Hamilton Park               40.727596   
57                  3210     Pershing Field               40.742677   
...                  ...                ...                     ...   
247442              3275     Columbus Drive               40.718355   
247443              3275     Columbus Drive               40.718355   
247537              3194    McGinley Square               40.725340   
247561              3195            Sip Ave               40.730743   
247580              3183     Exchange Place               40.716247   

        Start Station Longitude  End Station ID End Station Name  \
4                    -74.063784            3210   Pershing Field   
5                    -74.063784            3210   Pershing Field   
22                   -74.043991            3209     Brunswick St   
53                   -74.044247            3203    Hamilton Park   
57                   -74.051789            3195          Sip Ave   
...                         ...             ...              ...   
247442               -74.038914            3199     Newport Pkwy   
247443               -74.038914            3199     Newport Pkwy   
247537               -74.067622            3195          Sip Ave   
247561               -74.063784            3193     Lincoln Park   
247580               -74.033459            3183   Exchange Place   

        End Station Latitude  End Station Longitude  Bike ID   User Type  \
4                  40.742677             -74.051789    24573    Customer   
5                  40.742677             -74.051789    24442    Customer   
22                 40.724176             -74.050656    24662    Customer   
53                 40.727596             -74.044247    24444    Customer   
57                 40.730743             -74.063784    24573    Customer   
...                      ...                    ...      ...         ...   
247442             40.728745             -74.032108    24627    Customer   
247443             40.728745             -74.032108    26217    Customer   
247537             40.730743             -74.063784    24716  Subscriber   
247561             40.724605             -74.078406    24417  Subscriber   
247580             40.716247             -74.033459    24389    Customer   

        Birth Year  Gender  
4              NaN       0  
5              NaN       0  
22             NaN       0  
53             NaN       0  
57             NaN       0  
...            ...     ...  
247442         NaN       0  
247443         NaN       0  
247537         NaN       0  
247561         NaN       0  
247580         NaN       0  

[18999 rows x 15 columns]

## We notice from the snippet above that every 'Birth Year' of 'NaN' also has a 'Gender' of '0', so we will calculate the percentage, as suspected the percentage is 100%!

In [151]:
missing_birth_year_gender_zero = combined_df[(combined_df['Birth Year'].isnull()) & (combined_df['Gender'] == 0)]
count_missing_birth_gender_zero = len(missing_birth_year_gender_zero)
total_missing_birth_year = combined_df['Birth Year'].isnull().sum()
percentage_missing_birth_year_gender_zero = (count_missing_birth_gender_zero / total_missing_birth_year) * 100
percentage_missing_birth_year_gender_zero

100.0

## We also notice a lot of missing 'Birth Year' could be linked to the 'User Type' of 'Customer', as suspected, there is definitely a correlation.

In [152]:
missing_birth_year_by_user_type = combined_df[combined_df['Birth Year'].isnull()]['User Type'].value_counts()
total_missing_birth_year = combined_df['Birth Year'].isnull().sum()
percentage_missing_birth_year_by_user_type = (missing_birth_year_by_user_type / total_missing_birth_year) * 100
percentage_missing_birth_year_by_user_type

User Type
Customer      81.425338
Subscriber    18.574662
Name: count, dtype: float64

## We know out of the 18999 missing 'Birth Year' count that 81.42% of these are 'Customer' and 18.58% are 'Subscriber', we also know that all 18999 values will also have a 'Gender' of '0'

In [153]:
missing_birth_year_count = combined_df['Birth Year'].isnull().sum()
missing_birth_year_count

18999

## We see one 'Birth Year' of 1900, which is an outlier in our data, we remove this to get a more accurate representation

In [154]:
min_birth_year = combined_df['Birth Year'].nsmallest(10)
min_birth_year

215055    1900.0
136211    1934.0
164274    1937.0
180373    1937.0
187190    1937.0
213567    1937.0
112336    1940.0
149803    1940.0
177320    1940.0
1501      1941.0
Name: Birth Year, dtype: float64

In [155]:
combined_df.drop([215055], inplace=True)
combined_df

Trip Duration          Start Time           Stop Time  \
0                 362 2016-01-01 00:02:52 2016-01-01 00:08:54   
1                 200 2016-01-01 00:18:22 2016-01-01 00:21:42   
2                 202 2016-01-01 00:18:25 2016-01-01 00:21:47   
3                 248 2016-01-01 00:23:13 2016-01-01 00:27:21   
4                 903 2016-01-01 01:03:20 2016-01-01 01:18:24   
...               ...                 ...                 ...   
247579            557 2016-12-31 23:10:16 2016-12-31 23:19:33   
247580           2749 2016-12-31 23:29:39 2017-01-01 00:15:29   
247581            173 2016-12-31 23:44:37 2016-12-31 23:47:31   
247582           2424 2016-12-31 23:44:50 2017-01-01 00:25:14   
247583           2419 2016-12-31 23:44:50 2017-01-01 00:25:10   

        Start Station ID Start Station Name  Start Station Latitude  \
0                   3186      Grove St PATH               40.719586   
1                   3186      Grove St PATH               40.719586   
2                   3186      Grove St PATH               40.719586   
3                   3209       Brunswick St               40.724176   
4                   3195            Sip Ave               40.730743   
...                  ...                ...                     ...   
247579              3214   Essex Light Rail               40.712774   
247580              3183     Exchange Place               40.716247   
247581              3186      Grove St PATH               40.719586   
247582              3214   Essex Light Rail               40.712774   
247583              3214   Essex Light Rail               40.712774   

        Start Station Longitude  End Station ID  End Station Name  \
0                    -74.043117            3209      Brunswick St   
1                    -74.043117            3213    Van Vorst Park   
2                    -74.043117            3213    Van Vorst Park   
3                    -74.050656            3203     Hamilton Park   
4                    -74.063784            3210    Pershing Field   
...                         ...             ...               ...   
247579               -74.036486            3203     Hamilton Park   
247580               -74.033459            3183    Exchange Place   
247581               -74.043117            3270   Jersey & 6th St   
247582               -74.036486            3214  Essex Light Rail   
247583               -74.036486            3214  Essex Light Rail   

        End Station Latitude  End Station Longitude  Bike ID   User Type  \
0                  40.724176             -74.050656    24647  Subscriber   
1                  40.718489             -74.047727    24605  Subscriber   
2                  40.718489             -74.047727    24689  Subscriber   
3                  40.727596             -74.044247    24693  Subscriber   
4                  40.742677             -74.051789    24573    Customer   
...                      ...                    ...      ...         ...   
247579             40.727596             -74.044247    24465  Subscriber   
247580             40.716247             -74.033459    24389    Customer   
247581             40.725289             -74.045572    24641  Subscriber   
247582             40.712774             -74.036486    26219  Subscriber   
247583             40.712774             -74.036486    24471  Subscriber   

        Birth Year  Gender  
0           1964.0       2  
1           1962.0       1  
2           1962.0       2  
3           1984.0       1  
4              NaN       0  
...            ...     ...  
247579      1981.0       2  
247580         NaN       0  
247581      1978.0       1  
247582      1960.0       2  
247583      1956.0       1  

[247583 rows x 15 columns]

# Now, lets move on to the Weather data!

## Again, we will call .info() and check, we can see a Dtype that needs updating, and 2 columns that are entirely null which we can remove! We use axis=1 to specify the column, and inplace=True to keep the original DataFrame.

In [156]:
weather_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366 entries, 0 to 365
Data columns (total 16 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   STATION  366 non-null    object 
 1   NAME     366 non-null    object 
 2   DATE     366 non-null    object 
 3   AWND     366 non-null    float64
 4   PGTM     0 non-null      float64
 5   PRCP     366 non-null    float64
 6   SNOW     366 non-null    float64
 7   SNWD     366 non-null    float64
 8   TAVG     366 non-null    int64  
 9   TMAX     366 non-null    int64  
 10  TMIN     366 non-null    int64  
 11  TSUN     0 non-null      float64
 12  WDF2     366 non-null    int64  
 13  WDF5     364 non-null    float64
 14  WSF2     366 non-null    float64
 15  WSF5     364 non-null    float64
dtypes: float64(9), int64(4), object(3)
memory usage: 45.9+ KB


In [157]:
weather_df['DATE'] = pd.to_datetime(weather_df['DATE'])
weather_df.drop(['PGTM', 'TSUN'], axis=1, inplace=True)
weather_df


STATION                                         NAME       DATE  \
0    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-01   
1    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-02   
2    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-03   
3    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-04   
4    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-05   
..           ...                                          ...        ...   
361  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-27   
362  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-28   
363  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-29   
364  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-30   
365  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-31   

      AWND  PRCP  SNOW  SNWD  TAVG  TMAX  TMIN  WDF2   WDF5  WSF2  WSF5  
0    12.75  0.00   0.0   0.0    41    43    34   270  280.0  25.9  35.1  
1     9.40  0.00   0.0   0.0    36    42    30   260  260.0  21.0  25.1  
2    10.29  0.00   0.0   0.0    37    47    28   270  250.0  23.9  30.0  
3    17.22  0.00   0.0   0.0    32    35    14   330  330.0  25.9  33.1  
4     9.84  0.00   0.0   0.0    19    31    10   360  350.0  25.1  31.1  
..     ...   ...   ...   ...   ...   ...   ...   ...    ...   ...   ...  
361  13.65  0.01   0.0   0.0    53    62    40   270  270.0  29.1  38.0  
362   8.28  0.00   0.0   0.0    41    43    31   330  330.0  19.9  25.1  
363   8.05  0.36   0.0   0.0    38    45    31   170  150.0  18.1  25.1  
364  14.99  0.00   0.0   0.0    37    42    32   270  270.0  25.9  33.1  
365  12.30  0.00   0.0   0.0    35    44    29   200  220.0  21.9  28.0  

[366 rows x 14 columns]

## We check for duplicates, no duplicates are found which is good

In [158]:
weather_df.duplicated()


0      False
1      False
2      False
3      False
4      False
       ...  
361    False
362    False
363    False
364    False
365    False
Length: 366, dtype: bool

### Now, we use .describe(), the column names are quite confusing, lets rename some columns and drop some columns, the columns WDF2, WDF5, WSF2, WSF5 are wind speed and direction, the first column AWND (average wind) will provide enough information, so we will remove these 4 Columns. Also, Snowfall in inches is suffice, we do not need the snow depth as well

In [159]:
weather_df.describe()

DATE        AWND        PRCP        SNOW        SNWD  \
count                  366  366.000000  366.000000  366.000000  366.000000   
mean   2016-07-01 12:00:00    9.429973    0.104945    0.098087    0.342623   
min    2016-01-01 00:00:00    2.460000    0.000000    0.000000    0.000000   
25%    2016-04-01 06:00:00    6.765000    0.000000    0.000000    0.000000   
50%    2016-07-01 12:00:00    8.720000    0.000000    0.000000    0.000000   
75%    2016-09-30 18:00:00   11.410000    0.030000    0.000000    0.000000   
max    2016-12-31 00:00:00   22.820000    2.790000   24.000000   20.100000   
std                    NaN    3.748174    0.307496    1.276498    2.078510   

             TAVG        TMAX        TMIN        WDF2        WDF5        WSF2  \
count  366.000000  366.000000  366.000000  366.000000  364.000000  366.000000   
mean    57.196721   65.991803   48.459016  217.841530  228.269231   20.484426   
min      8.000000   18.000000    0.000000   10.000000   10.000000    6.900000   
25%     43.000000   51.250000   35.000000  150.000000  150.000000   15.000000   
50%     56.000000   66.000000   47.000000  240.000000  260.000000   19.900000   
75%     74.000000   83.000000   64.000000  300.000000  300.000000   23.900000   
max     89.000000   99.000000   80.000000  360.000000  360.000000   48.100000   
std     17.466981   18.606301   17.135790  102.548282   97.415777    6.848390   

             WSF5  
count  364.000000  
mean    26.801648  
min     10.100000  
25%     19.900000  
50%     25.100000  
75%     31.100000  
max     66.000000  
std      8.882610

In [160]:
weather_df.drop(['WDF2', 'WDF5', 'WSF2', 'WSF5', 'SNWD'], axis=1, inplace=True)
weather_df

STATION                                         NAME       DATE  \
0    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-01   
1    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-02   
2    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-03   
3    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-04   
4    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-05   
..           ...                                          ...        ...   
361  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-27   
362  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-28   
363  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-29   
364  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-30   
365  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-31   

      AWND  PRCP  SNOW  TAVG  TMAX  TMIN  
0    12.75  0.00   0.0    41    43    34  
1     9.40  0.00   0.0    36    42    30  
2    10.29  0.00   0.0    37    47    28  
3    17.22  0.00   0.0    32    35    14  
4     9.84  0.00   0.0    19    31    10  
..     ...   ...   ...   ...   ...   ...  
361  13.65  0.01   0.0    53    62    40  
362   8.28  0.00   0.0    41    43    31  
363   8.05  0.36   0.0    38    45    31  
364  14.99  0.00   0.0    37    42    32  
365  12.30  0.00   0.0    35    44    29  

[366 rows x 9 columns]

In [161]:
weather_df.rename(columns={'STATION': 'Station', 'NAME': 'Name', 'DATE': 'Date', 'SNOW': 'Snowfall_inch','TAVG': 'Average_Temp', 'PRCP': 'Precipitation_Inch', 'AWND': 'Average_Wind_MPH', 'TAVG': 'Average_Temp_C', 'TMAX': 'Max_Temp_C', 'TMIN': 'Min_Temp_C'}, inplace=True)
weather_df

Station                                         Name       Date  \
0    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-01   
1    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-02   
2    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-03   
3    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-04   
4    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-05   
..           ...                                          ...        ...   
361  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-27   
362  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-28   
363  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-29   
364  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-30   
365  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-31   

     Average_Wind_MPH  Precipitation_Inch  Snowfall_inch  Average_Temp_C  \
0               12.75                0.00            0.0              41   
1                9.40                0.00            0.0              36   
2               10.29                0.00            0.0              37   
3               17.22                0.00            0.0              32   
4                9.84                0.00            0.0              19   
..                ...                 ...            ...             ...   
361             13.65                0.01            0.0              53   
362              8.28                0.00            0.0              41   
363              8.05                0.36            0.0              38   
364             14.99                0.00            0.0              37   
365             12.30                0.00            0.0              35   

     Max_Temp_C  Min_Temp_C  
0            43          34  
1            42          30  
2            47          28  
3            35          14  
4            31          10  
..          ...         ...  
361          62          40  
362          43          31  
363          45          31  
364          42          32  
365          44          29  

[366 rows x 9 columns]

### We will now convert the temperatue from Fahrenheit into Celcius

In [162]:
weather_df['Average_Temp_C'] = (weather_df['Average_Temp_C'] - 32) * 5/9
weather_df['Max_Temp_C'] = (weather_df['Max_Temp_C'] - 32) * 5/9
weather_df['Min_Temp_C'] = (weather_df['Min_Temp_C'] - 32) * 5/9
weather_df

Station                                         Name       Date  \
0    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-01   
1    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-02   
2    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-03   
3    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-04   
4    USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-05   
..           ...                                          ...        ...   
361  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-27   
362  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-28   
363  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-29   
364  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-30   
365  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-12-31   

     Average_Wind_MPH  Precipitation_Inch  Snowfall_inch  Average_Temp_C  \
0               12.75                0.00            0.0        5.000000   
1                9.40                0.00            0.0        2.222222   
2               10.29                0.00            0.0        2.777778   
3               17.22                0.00            0.0        0.000000   
4                9.84                0.00            0.0       -7.222222   
..                ...                 ...            ...             ...   
361             13.65                0.01            0.0       11.666667   
362              8.28                0.00            0.0        5.000000   
363              8.05                0.36            0.0        3.333333   
364             14.99                0.00            0.0        2.777778   
365             12.30                0.00            0.0        1.666667   

     Max_Temp_C  Min_Temp_C  
0      6.111111    1.111111  
1      5.555556   -1.111111  
2      8.333333   -2.222222  
3      1.666667  -10.000000  
4     -0.555556  -12.222222  
..          ...         ...  
361   16.666667    4.444444  
362    6.111111   -0.555556  
363    7.222222   -0.555556  
364    5.555556    0.000000  
365    6.666667   -1.666667  

[366 rows x 9 columns]

## We have cleaned the data using pandas, the Database is set up in PGAdmin4 and we have succesfully connected the Database using SQLAlchemy, we will now create the Dataframes and insert Data into the Database

In [163]:

stations_df = combined_df[['Start Station ID', 'Start Station Name', 'Start Station Latitude', 'Start Station Longitude']].drop_duplicates()
stations_df.columns = ['Station_ID', 'Name', 'Latitude', 'Longitude']

users_df = combined_df[['Bike ID', 'User Type', 'Birth Year', 'Gender']].drop_duplicates()
users_df.columns = ['Bike_ID', 'User_Type', 'Birth_Year', 'Gender'] 

trips_df = combined_df[['Bike ID', 'Start Station ID', 'End Station ID', 'Start Time', 'Stop Time', 'Trip Duration', 'User Type', 'Birth Year', 'Gender']].drop_duplicates()
trips_df.columns = [['Bike_ID', 'Start_Station_ID', 'End_Station_ID', 'Start_Time', 'Stop_Time', 'Trip_Duration_Secs', 'User_Type', 'Birth_Year', 'Gender']]

bikes_df = combined_df[['Bike ID']].drop_duplicates()
bikes_df.columns = ['Bike_ID']  

weather_stations_df = weather_df[['Station', 'Name']].drop_duplicates()
weather_stations_df.columns = ['Weather_Station_ID', 'Name']  

weather_data_df = weather_df[['Station', 'Date', 'Average_Wind_MPH', 'Precipitation_Inch', 'Snowfall_inch', 'Average_Temp_C', 'Max_Temp_C', 'Min_Temp_C']]
weather_data_df.columns = ['Weather_Station_ID', 'Date', 'Average_Wind_MPH', 'Precipitation_Inch', 'Snowfall_Inch', 'Average_Temp_C', 'Max_Temp_C', 'Min_Temp_C']

## Below, we simply test that the new df's have been created correctly!

In [164]:
trips_df

Bike_ID Start_Station_ID End_Station_ID          Start_Time  \
0        24647             3186           3209 2016-01-01 00:02:52   
1        24605             3186           3213 2016-01-01 00:18:22   
2        24689             3186           3213 2016-01-01 00:18:25   
3        24693             3209           3203 2016-01-01 00:23:13   
4        24573             3195           3210 2016-01-01 01:03:20   
...        ...              ...            ...                 ...   
247579   24465             3214           3203 2016-12-31 23:10:16   
247580   24389             3183           3183 2016-12-31 23:29:39   
247581   24641             3186           3270 2016-12-31 23:44:37   
247582   26219             3214           3214 2016-12-31 23:44:50   
247583   24471             3214           3214 2016-12-31 23:44:50   

                 Stop_Time Trip_Duration_Secs   User_Type Birth_Year Gender  
0      2016-01-01 00:08:54                362  Subscriber     1964.0      2  
1      2016-01-01 00:21:42                200  Subscriber     1962.0      1  
2      2016-01-01 00:21:47                202  Subscriber     1962.0      2  
3      2016-01-01 00:27:21                248  Subscriber     1984.0      1  
4      2016-01-01 01:18:24                903    Customer        NaN      0  
...                    ...                ...         ...        ...    ...  
247579 2016-12-31 23:19:33                557  Subscriber     1981.0      2  
247580 2017-01-01 00:15:29               2749    Customer        NaN      0  
247581 2016-12-31 23:47:31                173  Subscriber     1978.0      1  
247582 2017-01-01 00:25:14               2424  Subscriber     1960.0      2  
247583 2017-01-01 00:25:10               2419  Subscriber     1956.0      1  

[247583 rows x 9 columns]

In [165]:
stations_df

Station_ID                     Name   Latitude  Longitude
0             3186            Grove St PATH  40.719586 -74.043117
3             3209             Brunswick St  40.724176 -74.050656
4             3195                  Sip Ave  40.730743 -74.063784
7             3211               Newark Ave  40.721525 -74.046305
8             3187                Warren St  40.721124 -74.038051
10            3183           Exchange Place  40.716247 -74.033459
11            3213           Van Vorst Park  40.718489 -74.047727
14            3193             Lincoln Park  40.724605 -74.078406
19            3194          McGinley Square  40.725340 -74.067622
20            3202             Newport PATH  40.727224 -74.033759
22            3196           Riverview Park  40.744319 -74.043991
27            3214         Essex Light Rail  40.712774 -74.036486
33            3207              Oakland Ave  40.737604 -74.052478
37            3199             Newport Pkwy  40.728745 -74.032108
40            3203            Hamilton Park  40.727596 -74.044247
41            3210           Pershing Field  40.742677 -74.051789
42            3190     Garfield Ave Station  40.710467 -74.070039
48            3185                City Hall  40.717732 -74.043845
55            3197                 North St  40.752559 -74.044725
61            3192       Liberty Light Rail  40.711242 -74.055701
62            3212          Christ Hospital  40.734786 -74.050444
65            3225    Baldwin at Montgomery  40.723659 -74.064194
73            3215              Central Ave  40.746730 -74.049251
101           3206                  Hilltop  40.731169 -74.057574
117           3184              Paulus Hook  40.714145 -74.033552
119           3205        JC Medical Center  40.716540 -74.049638
124           3198         Heights Elevator  40.748716 -74.040443
132           3220        5 Corners Library  40.734961 -74.059503
201           3191                 Union St  40.718211 -74.083639
203           3201                   Dey St  40.737711 -74.066921
930           3217             Bayside Park  40.698651 -74.082080
1033          3188                     NJCU  40.710109 -74.085849
2843          3200           MLK Light Rail  40.711131 -74.078885
4429          3216            Columbia Park  40.697030 -74.096937
5441          3189     West Side Light Rail  40.714402 -74.088772
104103        3270          Jersey & 6th St  40.725289 -74.045572
104299        3267             Morris Canal  40.712419 -74.038526
105982        3272             Jersey & 3rd  40.723332 -74.045953
106128        3268           Lafayette Park  40.713464 -74.062859
109539        3278         Monmouth and 6th  40.725685 -74.048790
109770        3279              Dixon Mills  40.721630 -74.049968
111266        3276         Marin Light Rail  40.714584 -74.042817
111398        3273             Manila & 1st  40.721651 -74.042884
111752        3275           Columbus Drive  40.718355 -74.038914
111943        3274           Bethune Center  40.704958 -74.085931
112102        3281      Leonard Gordon Park  40.745910 -74.057271
112414        3271      Danforth Light Rail  40.692640 -74.088012
207665        3269          Brunswick & 6th  40.726012 -74.050389
208641        3280              Astor Place  40.719282 -74.071262
211211        3426               JCBS Depot  40.709651 -74.068601
222168        3277  Communipaw & Berry Lane  40.714358 -74.066611

In [166]:
users_df

Bike_ID   User_Type  Birth_Year  Gender
0         24647  Subscriber      1964.0       2
1         24605  Subscriber      1962.0       1
2         24689  Subscriber      1962.0       2
3         24693  Subscriber      1984.0       1
4         24573    Customer         NaN       0
...         ...         ...         ...     ...
247516    26228  Subscriber      1988.0       2
247556    24641  Subscriber      1959.0       2
247557    24513  Subscriber      1967.0       2
247567    26241  Subscriber      1966.0       1
247582    26219  Subscriber      1960.0       2

[37486 rows x 4 columns]

In [167]:
bikes_df

Bike_ID
0         24647
1         24605
2         24689
3         24693
4         24573
...         ...
226120    22194
226722    27274
227084    18470
236263    18648
240399    18062

[566 rows x 1 columns]

In [168]:
weather_stations_df

Weather_Station_ID                                         Name
0        USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US

In [169]:
weather_data_df

Weather_Station_ID       Date  Average_Wind_MPH  Precipitation_Inch  \
0          USW00014734 2016-01-01             12.75                0.00   
1          USW00014734 2016-01-02              9.40                0.00   
2          USW00014734 2016-01-03             10.29                0.00   
3          USW00014734 2016-01-04             17.22                0.00   
4          USW00014734 2016-01-05              9.84                0.00   
..                 ...        ...               ...                 ...   
361        USW00014734 2016-12-27             13.65                0.01   
362        USW00014734 2016-12-28              8.28                0.00   
363        USW00014734 2016-12-29              8.05                0.36   
364        USW00014734 2016-12-30             14.99                0.00   
365        USW00014734 2016-12-31             12.30                0.00   

     Snowfall_Inch  Average_Temp_C  Max_Temp_C  Min_Temp_C  
0              0.0        5.000000    6.111111    1.111111  
1              0.0        2.222222    5.555556   -1.111111  
2              0.0        2.777778    8.333333   -2.222222  
3              0.0        0.000000    1.666667  -10.000000  
4              0.0       -7.222222   -0.555556  -12.222222  
..             ...             ...         ...         ...  
361            0.0       11.666667   16.666667    4.444444  
362            0.0        5.000000    6.111111   -0.555556  
363            0.0        3.333333    7.222222   -0.555556  
364            0.0        2.777778    5.555556    0.000000  
365            0.0        1.666667    6.666667   -1.666667  

[366 rows x 8 columns]

## Now, we will insert the data from these Dataframes into our pgAdmin4 database!

In [170]:
# Insert data into 'stations' table
stations_df.to_sql('stations', engine, if_exists='replace', index=False)

# Insert data into 'trips' table
trips_df.to_sql('trips', engine, if_exists='append', index=False)

# Insert data into 'bikes' table
bikes_df.to_sql('bikes', engine, if_exists='append', index=False)

# Insert data into 'users' table
users_df.to_sql('users', engine, if_exists='append', index=False, method='multi', chunksize=1000)

# Insert data into 'weather_stations' table
weather_stations_df.to_sql('weather_stations', engine, if_exists='append', index=False)

# Insert data into 'weather_data' table
weather_data_df.to_sql('weather_data', engine, if_exists='append', index=False)

366